test/extensions/filters/http/transformation/inja_transformer_replace_test.cc

```diff
@@ -0,0 +1,442 @@
+      const envoy::config::core::v3::Metadata *cluster_metadata) {
+  slot->set([](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
+          return std::make_shared<ThreadLocalTransformerContext>();
+  });
+  auto& typed_slot = slot->getTyped<ThreadLocalTransformerContext>();
+  typed_slot.header_map_ = &header_map;
+  typed_slot.request_headers_ = request_headers;
+  typed_slot.body_ = &body;
+  typed_slot.extractions_ = &extractions;
+  typed_slot.context_ = &context;
+  typed_slot.environ_ = &environ;
+  typed_slot.cluster_metadata_ = cluster_metadata;
+}
+
+TEST(Extraction, ExtractAndReplaceValueFromBodySubgroup) {
+  Http::TestRequestHeaderMapImpl headers{{":method", "GET"}, {":path", "/foo"}};
+
+  envoy::api::v2::filter::http::Extraction extractor;
+  extractor.mutable_body();
+  extractor.set_regex(".*(body)");
+  extractor.set_subgroup(1);
+  auto replacement_text = "BAZ";
+  extractor.mutable_replacement_text()->set_value(replacement_text);
+
+  NiceMock<Http::MockStreamDecoderFilterCallbacks> callbacks;
+  std::string body("not json body");
+  GetBodyFunc bodyfunc = [&body]() -> const std::string & { return body; };
+  std::string res(Extractor(extractor).extract(callbacks, headers, bodyfunc));
+
+  EXPECT_EQ("not json BAZ", res);
+}
+
+TEST(Extraction, ExtractAndReplaceValueFromFullBody) {
+  Http::TestRequestHeaderMapImpl headers{{":method", "GET"}, {":path", "/foo"}};
+
+  envoy::api::v2::filter::http::Extraction extractor;
+  extractor.mutable_body();
+  extractor.set_regex(".*");
+  extractor.set_subgroup(0);
+  auto replacement_text = "BAZ";
+  extractor.mutable_replacement_text()->set_value(replacement_text);
+
+  NiceMock<Http::MockStreamDecoderFilterCallbacks> callbacks;
+  std::string body("not json body");
+  GetBodyFunc bodyfunc = [&body]() -> const std::string & { return body; };
+  std::string res(Extractor(extractor).extract(callbacks, headers, bodyfunc));
+
+  EXPECT_EQ("BAZ", res);
+}
+
+// TODO: EDGE CASE !!!
+TEST(Extraction, ExtractAndReplaceAllFromFullBody) {
+  Http::TestRequestHeaderMapImpl headers{{":method", "GET"}, {":path", "/foo"}};
+
+  envoy::api::v2::filter::http::Extraction extractor;
+  extractor.mutable_body();
+  extractor.set_regex(".*");
+  extractor.set_subgroup(0);
+  extractor.set_replace_all(true);
+  auto replacement_text = "BAZ";
+  extractor.mutable_replacement_text()->set_value(replacement_text);
+
+  NiceMock<Http::MockStreamDecoderFilterCallbacks> callbacks;
+  std::string body("not json body");
+  GetBodyFunc bodyfunc = [&body]() -> const std::string & { return body; };
+  std::string res(Extractor(extractor).extract(callbacks, headers, bodyfunc));
+
+  // Note to self/reviewers: this is the current behavior, which is a kind of 
+  // confusing edge case in std::regex_replace when the regex is .*
+  // apparently, this regex matches the whole input string __AND__ the 
+  // line ending, so the replacement is applied twice
+  EXPECT_EQ("BAZBAZ", res);
```

Comment:
Note to reviewers: this edge case frustrates me, trying to figure out how to handle it at the moment

Reply:
does changing line 114 to `extractor.set_regex(".*$")` change anything? or maybe `"^.*$"`?